---
title: "BIO514"
description: |
  Bioinformatics - set up
site:
  distill::distill_article:
    toc: true
    toc_depth: 2
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Software

**REQUIRED** - We will be using RStudio to analyse the data set. It is recommend you have the following installed: [*RStudio version 1.4*](https://rstudio.com/products/rstudio/download/) or later and [*R version 4.0*](https://www.r-project.org/) or later. Further details on getting started in RStudio [here](http://siobhonlegan.com/research_site/rstudio/).


**Optional** (*not needed for today's workshop*) - We will not be doing the sequence pre-processing steps today but if you did want to do this you will need to download [conda](https://conda.io/projects/conda/en/latest/user-guide/install/index.html) and [QIIME2](https://qiime2.org/).

## Download data

1. **Raw Sequence** (*optional*)

Raw amplicon 16S sequence data from West et al. (2020) *Gut* 69, 1452-1459. doi: [10.1136/gutjnl-2019-319620](http://dx.doi.org/10.1136/gutjnl-2019-319620).
Download raw data from NCBI Sequence Read Archive. Project number `PRJNA493625` from https://sra-explorer.info/.

Rename so `fastq.gz` files in format `SAMPLEID_L001_R1_001.fastq.gz` (forward) or `SAMPLEID_L001_R2_001.fastq.gz` (reverse).

You will not be required to download this for today's tutorial but if you wanted you could use this data and follow the [sequence processing](04_sequence_processing.html) page.

2. **RData** (*REQUIRED*)

The easiest way to follow along with this tutorial is to download this GitHub repository using either option **1** or **2** below:

1. Go to https://github.com/siobhon-egan/BIO514 and click on the green **Code** button. Select **Download ZIP**, open/unzip the file. Open the `.Rmd` files in RStudio you will be able to follow along for the data analysis.


2. Use terminal and clone the GitHub repo.

```
git clone https://github.com/siobhon-egan/BIO514.git
```

